### PR TITLE
Include the kind of test while matching filters

### PR DESCRIPTION
**Pull request by sunshowers**
Consider a trial with name "my_test" and kind "foo". The text that gets printed out with `--list --format terse` is:

```
[foo] my_test: test
```

With libtest, an invariant is that the part of the string before ": test" can be passed into "--exact" to run that test. However, that invariant is currently not upheld by libtest-mimic: it only matches against the name of the test and not the kind. This breaks nextest's ability to run these tests: https://github.com/nextest-rs/nextest/issues/836

Make it so that libtest-mimic also matches against the kind of test, not just its name.

I also took the liberty of running rustfmt (or really, my editor automatically ran it when I went in to edit the file.)

### CODE REVIEW

File: tests/mixed_bag.rs [hunk 1]
```diff
@@ -227,11 +227,32 @@ fn list_ignored() {
...
+        num_failed: 0,
+        num_ignored: 0,
+        num_measured: 0,
+     });
+}
+
+#[test]
+fn list_with_filter_exact() {
+    let (c, out) = common::do_run(args(["--list", "--exact", "[apple] fox"]), tests());
+    // Matches all tests that contain "a" in either the name or the kind.
```
Comment (niluxv) [inline tests/mixed_bag.rs]: I think this comment is left over from copy-paste, and not accurate here...

File: src/lib.rs [hunk 2]
```diff
@@ -364,22 +371,28 @@ impl Arguments {
...
+        //
+        // then "--exact <some string>" runs exactly that test.
+        let test_name_with_kind = test.info.test_name_with_kind();
 
         // If a filter was specified, apply this
         if let Some(filter) = &self.filter {
             match self.exact {
-                true if test_name != filter => return true,
-                false if !test_name.contains(filter) => return true,
+                true if &test_name_with_kind != filter => return true,
```
Comment (LukasKalbertodt) [inline src/lib.rs]: This is too much of a breaking behavior change as I'm willing to release as a minor version :/
Further, I can totally see people not wanting to specify the kind, but only run `--exact foo::bar` or something like that. What do you think about `&test_name_with_kind != filter && test_name != filter`? I.e. if `--exact` is specified, you can either specify only the test name OR the test name with kind?

File: src/lib.rs [hunk 3]
```diff
@@ -364,22 +371,28 @@ impl Arguments {
...
                 _ => {}
             };
         }
 
         // If any skip pattern were specified, test for all patterns.
         for skip_filter in &self.skip {
             match self.exact {
-                true if test_name == skip_filter => return true,
-                false if test_name.contains(skip_filter) => return true,
+                true if &test_name_with_kind == skip_filter => return true,
```
Comment (LukasKalbertodt) [inline src/lib.rs]: Same here

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
Thanks for the PR. Seems like a reasonable idea to also match the kind.

I left one comment that still worries me. And please remove the rustfmt changes so that this PR only contains the changes for the behavior change. Plus the copy&paste comment that the other reviewer found.

File: src/lib.rs (hunk 2 above)
Comment (sunshowers) [reply src/lib.rs]: Addressed this in the update.

File: src/lib.rs (hunk 3 above)
Comment (sunshowers) [reply src/lib.rs]: Addressed this as well.

**Comment by sunshowers** [comment]
> And please remove the rustfmt changes so that this PR only contains the changes for the behavior change

Would you be open at all to formatting the code via rustfmt, maybe as a separate change? To be honest this is a huge discouragement for me to contribute (and why I didn't get around to this for many months), because my editor is set up to autoformat on save, and rustfmt is used by every single one of the ~100 Rust projects I've ever contributed to.

File: tests/mixed_bag.rs (hunk 1 above)
Comment (sunshowers) [reply tests/mixed_bag.rs]: Addressed this, thanks.

**Review by LukasKalbertodt — APPROVED** [review]
Thank you! Also for writing tests, that's always very helpful.

**Comment by LukasKalbertodt** [comment]
> Would you be open at all to formatting the code via rustfmt, maybe as a separate change? To be honest this is a huge discouragement for me to contribute (and why I didn't get around to this for many months), because my editor is set up to autoformat on save, and rustfmt is used by every single one of the ~100 Rust projects I've ever contributed to.

Unfortunately, not at this time, no. I just checked all the changes made by `cargo fmt` again. And while most are totally fine, I disagree with some. My main concern [I raised in an issue](https://github.com/rust-lang/rustfmt/issues/3381#issuecomment-573777732) already (and quite a few people seem to agree), but changing the default was rejected and the configuration is still not stable. Another thing is that `rustfmt` removes trailing commas in attributes, which makes no sense to me. And it doesn't seem like there is a configuration option for this at all. I would love to use rustfmt, but there are some blockers for me. Of course, once a project grows and gets more contributors, I might change my opinion for that particular project.

And FWIW, my editor is also set up to do certain things automatically on save (removing trailing whitespace, ...). But if I contribute to some project that does not adhere to this basic code hygiene, I will disable it temporarily. Or more likely, just don't stage those changes before a commit. Personally I think that's no big deal. 

That's just some explanation why things are the way they are.

**Comment by LukasKalbertodt** [comment]
Released as v0.7.0

**Pull request merged**